### ISSUE TITLE
Pull Quote block displays several background classes at the same time

### ISSUE THREAD

**Issue by dashkevych** [WordPress/gutenberg#11352]
**Describe the bug**
When custom color palette is added via `add_theme_support( 'editor-color-palette')`, and Pull Quote (style: solid color) block is used, the element class attribute contains multiple background classes.

Here is a screenshot -- front end: https://snag.gy/C3sY2o.jpg
Here is a screenshot -- dashboard: https://snag.gy/syqOQ5.jpg

**To Reproduce**
1. Add Pullquote block with a solid color style to the editor.
2. Select the 1st Main Color in the Color Settings box.
3. Update the post.
4. Change your Main Color to the 2nd color in the Color Settings box.
5. Update the post.
6. Change your Main Color to the 3nd color in the Color Settings box.
7. Update the post.

**Expected behavior**
Pullquote block should have only one background class.

**Desktop (please complete the following information):**
 - OS: macOS 10.13.6
 - Chrome
 - Version 69.0.3497.100
 - Gutenberg: Version 4.1.1

**Comment by ocean90** [comment]
Good catch! It's a bit tricky to reproduce, I have to do a page reload before changing the color so I get the multiple CSS classes.

I think the root for this is in the custom class name filters and [`addParsedDifference()`](https://github.com/WordPress/gutenberg/blob/7c7e764c5f045f7dcf7730ca8bba3df28bfd46de/packages/editor/src/hooks/custom-class-name.js#L121-L153).

**Comment by ocean90** [comment]
@aduth Can you confirm that the issue lies in addParsedDifference()?

**Comment by aduth** [comment]
Initial impressions:

@youknowriad Is there a reason we don't assign the selected style as an attribute? I might understand if the idea was that there should only be stylistic effects in changing the selected style, but in practice one of the causes of the issue here is that we treat it as a data point in the implementation of the Pullquote block in deciding how to construct the saved output markup:

https://github.com/WordPress/gutenberg/blob/cb585754f92beee040af6da3e394916f0255f85b/packages/block-library/src/pullquote/index.js#L80-L85

I'd think we should either want to (a) treat it as a true data point in the form of an attribute or (b) never allow ourselves to consider it in the saved markup of a block (cc @jorgefilipecosta).

I'm still investigating `addParsedDifference`, in particular this line which is the likely culprit for Pullquote, in that the `className` containing `is-style-` won't reach the subsequent `save` logic:

https://github.com/WordPress/gutenberg/blob/7c7e764c5f045f7dcf7730ca8bba3df28bfd46de/packages/editor/src/hooks/custom-class-name.js#L139

**Comment by aduth** [comment]
The need to omit the `className` attribute in generating the default saved result was introduced in #8232 to resolve an invalidation which could occur when markup was manually changed to drop a custom class name (cc @johngodley).

**Comment by aduth** [comment]
Pull request at #11828

**Comment by youknowriad** [comment]
@aduth I think the reason it's reusing 'className' instead of a custom attribute is because it allows us to take advantage of the specific handling of extra classNames. Which means if style variations are defined or not, it's not important, the className is kept, if it's added manually, it works as well without validation warnings etc...

I considered adding a new prop: `activeStyle` to `edit`/`save` functions which is kind of redundant with `className` but it prevents the block author from performing the `getActiveStyle` call manually.

**Comment by carolinan** [comment]
Since the solid color block style has been removed, and the block supports background, text, link and border colors, I am closing this issue as solved. I do not see any duplicate class names in the editor or front.
Please reopen it (or ask me to reopen it) if you think this is incorrect.
